groups/bsl/bslstl/bslstl_variant.h

```diff
@@ -0,0 +1,3338 @@
+struct Variant_ConvertToIndexImpl;
+
+template <size_t INDEX, class TYPE>
+struct Variant_ConvertToIndexImpl<INDEX, TYPE, VARIANT_NOT_A_TYPE>
+: bsl::integral_constant<size_t, bsl::variant_npos> {
+    // Primary template is used when ARGS is empty, i.e. the type wasn't found
+    // in the given list of template type arguments.
+};
+template <size_t INDEX, class TYPE, class HEAD, class... TAIL>
+struct Variant_ConvertToIndexImpl<INDEX, TYPE, HEAD, TAIL...>
+: bsl::conditional<bsl::is_same<typename bsl::remove_cvref<TYPE>::type,
+                                typename bsl::remove_cvref<HEAD>::type>::value,
+                   bsl::integral_constant<size_t, INDEX>,
+                   Variant_TypeToIndexImpl<INDEX + 1, TYPE, TAIL...> >::type {
+    // If the current 'HEAD' type of the list of template type arguments is the
+    // same as the 'TYPE' to find, the value of this metafunction is the zero
+    // based index of the 'HEAD' in the original list of template type
+    // arguments.  Otherwise, we check the 'TAIL..' for 'TYPE'.
+};
+
+template <class TYPE, class VARIANT>
+struct Variant_ConvertIndex;  // primary template not defined
+
+template <class TYPE, class HEAD, class... TAIL>
+struct Variant_ConvertIndex<TYPE, bsl::variant<HEAD, TAIL...> >
+: Variant_ConvertToIndexImpl<
+      0,
+      typename bslmf::MovableRefUtil::RemoveReference<TYPE>::type,
+      HEAD,
+      TAIL...> {
+    // This metafunction calculates the zero-based index of 'TYPE' in a list of
+    // variadic type template arguments.  If 'TYPE' does not appear in the list
+    // of variadic template type arguments, the value of this metafunction is
+    // 'bsl::variant_npos'.  Used in assignment or construction from a 'TYPE'
+    // and does not require that the specified 'TYPE' matches the CV
+    // qualifications of the alternative type.
+};
+
+#endif
+#endif  //BSL_VARIANT_FULL_IMPLEMENTATION
+
+template <class VARIANT, class TYPE>
+struct Variant_ConstructsFromType
+: bsl::integral_constant<
+      bool,
+      !bsl::is_same<typename bsl::remove_cvref<TYPE>::type, VARIANT>::value &&
+          !Variant_IsTag<typename bsl::remove_cvref<TYPE>::type>::value &&
+          BSLSTL_VARIANT_CONVERT_INDEX_OF(TYPE, VARIANT) !=
+              bsl::variant_npos> {
+    // This metafunction is derived from 'bsl::true_type' if 'TYPE' is not a
+    // tag type or the 'variant' type, and if there's unique alternative that
+    // is the best conversion match from 'TYPE'.
+};
+
+template <class VARIANT, class TYPE>
+struct Variant_AssignsFromType
+: bsl::integral_constant<
+      bool,
+      !bsl::is_same<typename bsl::remove_cvref<TYPE>::type, VARIANT>::value &&
+          BSLSTL_VARIANT_CONVERT_INDEX_OF(TYPE, VARIANT) !=
+              bsl::variant_npos> {
+    // This metafunction is derived from 'bsl::true_type' if 'TYPE' is not the
+    // 'variant' type, and if there's unique alternative that is the best
+    // conversion match from 'TYPE'.
+};
```

Comment:
Sort of, of kind of, but it has a slightly different problem. 
BSLSTL_VARIANT_CONVERT_INDEX_OF requires copy constructibility of the selected alternative, but only for the selected alternative. This isn't obvious and probably needs documenting,
The implementation will not compile if the selected alternative is not assignable(be it const alternative or an alternative that doesn't support assignment). 
std::variant requires all alternatives to be assignable and copy constructible. For C++11 we can add those requirements. For C++03, the assignment of all alternatives will be difficult to check, but perhaps we can do some trickery which doesn't SFINAE but also doesn't compile if one attempts an assignment to a variant with a non-assignable alternative.

Reply:
Sort of, of kind of, but it has a slightly different problem.

The implementation will not compile if the selected alternative is not assignable(be it const alternative or an alternative that doesn't support assignment) or because it isn't constructible. However, it will participate in the overload resolution, while in the standard it is SFINAE-ed away. Annoyingly, my negative tests only tested for the failure, but I didn't consider what the failure was.

We have a similar problem with a few other constructors. 

I'll add constraints where needed. The assignable trait isn't available in C++03, so we won't always be able to have identical behaviour between C++03 and C++11, but that's already the case for anything that requires assignability in optional

Reply:
I've added the explicit is_constructible and is_assignable constraints within `Variant_AssignsFromType` and `Variant_ConstructsFromType`. We do now get SFINAE in non special member functions, where we would have had compile errors before. The implementation is still limited in much the same way when it comes to special member functions.  A possible improvement would be to have a C++11 only change to add base classes which do the copy/move construction and copy/move assignment, and then partially specialise those on whether an operation is available in all the alternatives (provide a special member function if needed, delete it in the specialisation for when not all alternatives support the special member function). I'm going to say that's probably adding even more complexity to the difference between C++03/C++11, but it should be considered if std::variant backed C++17 implementation is ever made.

Reply:
When i say special member functions, I mean the non templated default, copy, and move constructor. They are not constrained so they don't get SFINAE-d away. At the moment, these are preferred over all the templated constructors.  Forcing them into a template would mess up the overload resolution. 

What gcc does is have a hierarchy of base classes for special member functions. Then they have a specialisation of each base class which provides a defaulted (and consequently deleted) special member function, depending on the alternatives. That way, the overload set is correct, and if a special member function is selected over a templated constructor, you get an error rather than a SFINAE. It also benefits cases when the selected construction would be trivial, as you can get triviality of the variant too. This is what I mean by having such a base class hierarchy as a possible improvement in the future.

Reply:
Right, the thing I'm confused about is not the meaning of "special member function" but the thing you said earlier about how the "majority of cases that would have fallen into the new constrained constructor/assignment select a special member function".

I guess it depends on what you perceive as the "majority". If you try something like

```
variant<const char*, NonCopyable> v1, v2;
v1 = v2;
```

then in the old version, this would have selected the `operator=(T&&)` and there would be a compile error because none of the alternatives can be constructed from or assigned to from `v2`. In the new version, this selects the copy assignment operator and we just get a different compile error.

If you have something like

```
variant<const char*, NonCopyable> v;
v = 1;
```

then in the old version it would select `operator=(T&&)` and give a compile error, and in the new version there is no viable candidate; the copy assignment operator isn't viable. So if this `v = 1` is in an SFINAE context then you actually get a substitution failure in this case, which may be more desirable.

If you're saying that the first scenario is the majority of cases then I guess we can say that in the majority of cases there's no improvement (hard error before, hard error after). I thought of the second scenario as the majority, so I was confused. Anyway, I don't think it really matters. Did I understand everything correctly this time?